### ISSUE TITLE
Large differences observed in inference results between BERT model quantized with Fastertransformer QAT int8-mode2 and unquantized version.

### ISSUE THREAD

**Issue by alextsao1999** [NVIDIA/FasterTransformer#488]
I recently used the Fastertransformer QAT int8-mode-2 quantization method to quantize a BERT model, and saved the quantized weights to /workspace/data/bert-base-uncased-QAT-mode-2-KD. However, when I tested the quantized version against the unquantized version using examples/pytorch/bert/bert_example.py, I observed significant differences between the two versions. I am wondering if there is something wrong with my implementation or if there is a known issue with Fastertransformer QAT-mode-2.

Steps to reproduce:
1. cd examples/pytorch/bert/bert-quantization-sparsity
2. python run_squad.py \
  --init_checkpoint=$MODEL_DIR/bert-base-uncased-finetuned/pytorch_model.bin \
  --do_calib \
  --train_file=$DATA_DIR/v1.1/train-v1.1.json \
  --train_batch_size=16 \
  --num-calib-batch=16 \
  --do_lower_case \
  --bert_model=bert-base-uncased \
  --max_seq_length=384 \
  --doc_stride=128 \
  --vocab_file=$MODEL_DIR/bert-base-uncased/vocab.txt \
  --config_file=$MODEL_DIR/bert-base-uncased/config.json \
  --json-summary=$MODEL_DIR/dllogger.json \
  --output_dir=$MODEL_DIR/bert-base-uncased-PTQ-mode-2-for-KD \
  --max_steps=-1 \
  --fp16 \
  --calibrator percentile \
  --percentile 99.99 \
  --quant_mode ft2
3. python -m torch.distributed.launch --nproc_per_node=8 run_squad.py \
  --init_checkpoint=$MODEL_DIR/bert-base-uncased-PTQ-mode-2-for-KD/pytorch_model.bin \
  --do_train \
  --train_file=$DATA_DIR/v1.1/train-v1.1.json \
  --train_batch_size=4 \
  --learning_rate=3e-5 \
  --num_train_epochs=10 \
  --do_predict \
  --predict_file=$DATA_DIR/v1.1/dev-v1.1.json \
  --eval_script=$DATA_DIR/v1.1/evaluate-v1.1.py \
  --do_eval \
  --do_lower_case \
  --bert_model=bert-base-uncased \
  --max_seq_length=384 \
  --doc_stride=128 \
  --vocab_file=$MODEL_DIR/bert-base-uncased/vocab.txt \
  --config_file=$MODEL_DIR/bert-base-uncased/config.json \
  --json-summary=$MODEL_DIR/dllogger.json \
  --output_dir=$MODEL_DIR/bert-base-uncased-QAT-mode-2-KD \
  --max_steps=-1 \
  --fp16 \
  --quant_mode ft2 \
  --distillation \
  --teacher=$MODEL_DIR/bert-base-uncased-finetuned/pytorch_model.bin
4. Comment the line bert_example.py:166 -> # ft_weights._generated_weights = True  # for int8 handling
5. python bert_example.py 16 12 384 12 64 --time --data_type=fp16 --int8_mode=2 --weight_path=/workspace/data/bert-base-uncased-PTQ-mode-2-for-KD/pytorch_model.bin

Results: 
Data range:  [-2.076171875, 5.6171875]
FT Data range:  [-1.4716796875, 10.5625]
FT Mean diff: 0.2657986879348755
FT Max diff:  5.1171875
FT Min diff:  0.0
EFF-FT Data range:  [-1.4716796875, 10.5625]
EFF-FT Mean diff: 0.2657986879348755
EFF-FT Max diff:  5.1171875
EFF-FT Min diff:  0.0

Additional information:
The BERT model was properly quantized without any errors.
The same test data was used for both the quantized and unquantized versions.
Please let me know if there is anything I can provide to help diagnose the issue. Thank you for your help.

**Comment by byshiue** [comment]
Hi, @alextsao1999. Because the baseline is torch with FP16, but not int8. It is often to observe large gap in such case. Can you evaluate the checkpoint on squad directly?

**Comment by alextsao1999** [comment]
Thank you for your response. Are you suggesting that fp16 and int8 are not comparable? I was simply curious about the extent to which the accuracy would decrease when using qat int8 compared to fp16. Any further insights would be greatly appreciated.

**Comment by byshiue** [comment]
> Thank you for your response. Are you suggesting that fp16 and int8 are not comparable? I was simply curious about the extent to which the accuracy would decrease when using qat int8 compared to fp16. Any further insights would be greatly appreciated.

We cannot compare the values of tensor of fp16 and int8 directly, because the target of QAT is not minimizing their diff, but maximize the accuracy of real task. So, you should evaluate the int8 QAT on SQuAD or other real tasks directly. From our experiments, int8 QAT can keep same accuracy as fp16.

**Comment by alextsao1999** [comment]
> > Thank you for your response. Are you suggesting that fp16 and int8 are not comparable? I was simply curious about the extent to which the accuracy would decrease when using qat int8 compared to fp16. Any further insights would be greatly appreciated.
> 
> We cannot compare the values of tensor of fp16 and int8 directly, because the target of QAT is not minimizing their diff, but maximize the accuracy of real task. So, you should evaluate the int8 QAT on SQuAD or other real tasks directly. From our experiments, int8 QAT can keep same accuracy as fp16.

When I evaluated the checkpoint on SQuAD, I didn't seem to find any calls related to FasterTransformer. I was just curious if the results of the evaluation could be aligned with FasterTransformer's Bert model results. Can you please help me understand why the evaluated checkpoint on SQuAD can achieve the same accuracy as FasterTransformer's fp16 implementation?

**Comment by byshiue** [comment]
> Can you please help me understand why the evaluated checkpoint on SQuAD can achieve the same accuracy as FasterTransformer's fp16 int8-qat implementation?

Can you explain more? The reason int8-qat can achieve same accuracy as fp16 is because QAT. So, I don't know how to explain.
